controllers/storageoscluster/metrics-exporter_operand.go

```diff
@@ -0,0 +1,195 @@
+
+// dependenciesFound dictates wether the dependencies are present
+// and thus we can apply the metrics-exporter resources
+// currently we depend on Prometheus CRDs
+func (c *MetricsExporterOperand) dependenciesFound() bool {
+	return c.prometheusCRDsInstalled
+}
+
+func (c *MetricsExporterOperand) updateDependenciesState(currentState bool) {
+	c.prometheusCRDsLock <- struct{}{}
+	c.prometheusCRDsInstalled = currentState
+	<-c.prometheusCRDsLock
+}
+
+func (c *MetricsExporterOperand) ReadyCheck(ctx context.Context, obj client.Object) (bool, error) {
+	ctx, span, _, log := instrumentation.Start(ctx, "MetricsExporterOperand.ReadyCheck")
+	defer span.End()
+
+	ctx, cancel := context.WithTimeout(ctx, time.Minute)
+	defer cancel()
+
+	if !c.dependenciesFound() {
+		return false, nil
+	}
+
+	nodeDS := &appsv1.DaemonSet{}
+	key := client.ObjectKey{Name: "storageos-node", Namespace: obj.GetNamespace()}
+	if err := c.client.Get(ctx, key, nodeDS); err != nil {
+		return false, err
+	}
+
+	metricsDS := &appsv1.DaemonSet{}
+	key = client.ObjectKey{Name: "storageos-metrics-exporter", Namespace: obj.GetNamespace()}
+	if err := c.client.Get(ctx, key, metricsDS); err != nil {
+		return false, err
+	}
+
+	// metrics-exporter only returns complete metrics when there's one instance running
+	// alongside each node container. Failing to do so means the metrics can be missing
+	// data for some volumes.
+	if metricsDS.Status.NumberReady == nodeDS.Status.DesiredNumberScheduled {
+		log.Info("metrics-exporter is ready", "NumberReady", metricsDS.Status.NumberReady)
+		return true, nil
+	}
+
+	log.Info("metrics-exporter not ready")
+	return false, nil
+}
+
+func (c *MetricsExporterOperand) Ensure(ctx context.Context, obj client.Object, ownerRef metav1.OwnerReference) (eventv1.ReconcilerEvent, error) {
+	ctx, span, _, log := instrumentation.Start(ctx, "MetricsExporterOperand.Ensure")
+	defer span.End()
+
+	b, err := getMetricsExporterBuilder(c.fs, obj, c.kubectlClient)
+	if err != nil {
+		span.RecordError(err)
+		return nil, err
+	}
+
+	err = b.Apply(ctx)
+	if err != nil {
+		// the error type is lost along the way so we check against its val
+
+		// if the CRD is not present before installing metrics-exporter
+		if strings.Contains(err.Error(), "no matches for kind \"ServiceMonitor\" in version \"monitoring.coreos.com/v1\"") ||
+			// if CRD is deleted after having installed metrics-exporter
+			strings.Contains(err.Error(), "the server could not find the requested resource (post servicemonitors.monitoring.coreos.com)") {
+			log.Info("prometheus CRDs not found, metrics-exporter won't be installed")
+
+			// prometheus CRDs not found, abort and update state
+			c.updateDependenciesState(false)
+
+			return nil, nil
+		}
+
+		span.RecordError(err)
+		return nil, err
+	}
+
+	c.updateDependenciesState(true)
+
+	// TODO: Apply only at creation. Subsequent updates should only update
```

Comment:
This TODO should have a ticket associated with it

Reply:
It's an old todo shared amongst all operands and it's here to due to copy and paste
I don't have enough knowledge as to why we should/shouldn't do this to raise a ticket but didn't want to remove it either